crypto/bio/bio_lib.c

```diff
@@ -786,53 +786,62 @@ void bio_cleanup(void)
 }
 
 /* Internal variant of the below BIO_wait() not calling BIOerr() */
-static int bio_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
+static int bio_wait(BIO *bio, time_t max_time, unsigned int polling_millisec)
 {
 #ifndef OPENSSL_NO_SOCK
     int fd;
 #endif
+    long sec_diff;
 
-    if (max_time == 0)
+    if (max_time == 0) /* no timeout */
         return 1;
 
 #ifndef OPENSSL_NO_SOCK
-    if (BIO_get_fd(bio, &fd) > 0)
+    if (BIO_get_fd(bio, &fd) > 0 && fd < FD_SETSIZE)
         return BIO_socket_wait(fd, BIO_should_read(bio), max_time);
 #endif
-    if (milliseconds > 1000) {
-        long sec_diff = (long)(max_time - time(NULL)); /* might overflow */
+    /* fall back to polling since no sockets are available */
 
-        if (sec_diff <= 0)
-            return 0; /* timeout */
-        if ((unsigned long)sec_diff < milliseconds / 1000)
-            milliseconds = (unsigned long)sec_diff * 1000;
+    sec_diff = (long)(max_time - time(NULL)); /* might overflow */
+    if (sec_diff < 0)
```

Comment:
previously this was if (sec_diff <= 0) was this change intentional?

Reply:
Yes, because if `sec_diff == 0` the calculation of the minimum would go wrong and the sec_diff resolution insufficient if `polling_millisec` < 1000.

Reply:
I don't understand your reasoning, as you say `if sec_diff == 0 the calculation of the minimum would go wrong`
I think you never want a sleep of 0 ms.  The easiest fix would be to consider sec_diff == 0 as time out.

Reply:
By `would go wrong` if `sec_diff == 0` I meant that in this case `min(sec_diff * 1000, nap_milliseconds)` would be 0 ms, and this is what the code was intended to prevent. 
But the calculation was really buggy - I've just revised it.
Thanks for having a close look at that!